File: 2024/Day06_2.cpp
// This version implemented by Cursor based on Day06.java
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_map>
#include <chrono>
#include <string>
#include <bitset>
#include <thread>
#include <mutex>
#include <memory>

// Cache-friendly compact structures with minimal memory footprint
struct Coord {
    int16_t row, col;
    
    Coord(int16_t r, int16_t c) : row(r), col(c) {}
    
    bool operator==(const Coord& other) const {
        return row == other.row && col == other.col;
    }
};

// Custom hash function for Coord - optimized bit manipulation
namespace std {
    template <>
    struct hash<Coord> {
        size_t operator()(const Coord& c) const {
            // Faster than shifting when coordinate range is limited
            return (c.row * 10000) + c.col;
        }
    };
}

// Efficient guard structure
struct Guard {
    int16_t row, col;
    char dir;
    
    Guard(int16_t r, int16_t c, char d) : row(r), col(c), dir(d) {}
};

const std::string filepath = "inputs/day06.txt";

// Direction constants and lookup tables - aligned for cache efficiency
static constexpr int8_t DIR_UP = 0, DIR_DOWN = 1, DIR_LEFT = 2, DIR_RIGHT = 3;
static constexpr int8_t dirRow[4] = {-1, 1, 0, 0};   // ^ v < >
static constexpr int8_t dirCol[4] = {0, 0, -1, 1};   // ^ v < >
static constexpr char dirChars[4] = {'^', 'v', '<', '>'};
static constexpr char turnMap[4] = {'>', '<', '^', 'v'}; // ^ v < >

// We use strings instead of vector<char> for grid - much more efficient for this case
using Grid = std::vector<std::string>;
using PathMap = std::unordered_map<Coord, std::bitset<4>>;

// Pre-computed fast lookup for direction index
static int8_t fastDirIndex[256];

// Initialize lookup tables
__attribute__((constructor))
static void initTables() {
    // Set all to -1 by default
    for (int i = 0; i < 256; i++) {
        fastDirIndex[i] = -1;
    }
    
    // Set the direction values
    fastDirIndex[static_cast<unsigned char>('^')] = DIR_UP;
    fastDirIndex[static_cast<unsigned char>('v')] = DIR_DOWN;
    fastDirIndex[static_cast<unsigned char>('<')] = DIR_LEFT;
    fastDirIndex[static_cast<unsigned char>('>')] = DIR_RIGHT;
}

// Super-fast inline direction lookup
inline int8_t getDirIndex(char dir) {
    return fastDirIndex[static_cast<unsigned char>(dir)];
}

inline char getTurn(char dir) {
    return turnMap[getDirIndex(dir)];
}

// Load grid once and reuse - with memory pre-allocation
Grid getGrid() {
    Grid grid;
    std::ifstream file(filepath);
    
    if (!file.is_open()) {
        std::cerr << "Error opening file: " << filepath << std::endl;
        return grid;
    }
    
    grid.reserve(200); // Pre-allocate (adjusted based on expected size)
    std::string line;
    // Count file lines first to optimize allocation
    while (std::getline(file, line)) {
        grid.push_back(std::move(line));
    }
    
    return grid;
}

// Fast guard position finder - optimized with early return
Guard findGuard(const Grid& grid) {
    for (int16_t i = 0; i < grid.size(); i++) {
        const std::string& row = grid[i];
        for (int16_t j = 0; j < row.length(); j++) {
            if (row[j] == '^') {
                return Guard(i, j, '^');
            }
        }
    }
    return Guard(0, 0, '^');
}

// Fast boundary check
inline bool isInGrid(const Grid& grid, int16_t row, int16_t col) {
    return row >= 0 && row < grid.size() && col >= 0 && col < grid[row].length();
}

void moveGuard(const Grid& grid, Guard& guard) {
    int8_t dirIdx = getDirIndex(guard.dir);
    int16_t nextRow = guard.row + dirRow[dirIdx];
    int16_t nextCol = guard.col + dirCol[dirIdx];
    
    if (isInGrid(grid, nextRow, nextCol) && grid[nextRow][nextCol] == '#') {
        guard.dir = turnMap[dirIdx];
    } else {
        guard.row = nextRow;
        guard.col = nextCol;
    }
}

// Optimized path update - faster hash map lookups
bool updatePath(PathMap& path, const Guard& guard) {
    Coord pos(guard.row, guard.col);
    int8_t dirBit = getDirIndex(guard.dir);
    
    auto it = path.find(pos);
    if (it != path.end()) {
        // If we've been here in this direction before, cycle detected
        if (it->second.test(dirBit)) {
            return false;
        }
        it->second.set(dirBit);
    } else {
        std::bitset<4> dirs;
        dirs.set(dirBit);
        path.emplace(pos, dirs);
    }
    
    return true;
}

// Optimized solver
PathMap solve(const Grid& grid) {
    Guard guard = findGuard(grid);
    PathMap path;
    path.reserve(5000); // Pre-allocate hash map size
    
    // Initialize starting position with UP direction
    path[Coord(guard.row, guard.col)].set(DIR_UP);
    
    // Specialized fast path for movement
    while (true) {
        moveGuard(grid, guard);
        
        if (!isInGrid(grid, guard.row, guard.col)) {
            break;
        }
        
        if (!updatePath(path, guard)) {
            return PathMap();
        }
    }
    
    return path;
}

PathMap runPt1(const Grid& grid) {
    return solve(grid);
}

// Multi-threaded part 2 solution
int runPt2(const Grid& originalGrid, const PathMap& pt1Path) {
    int count = 0;
    Guard startGuard = findGuard(originalGrid);
    
    // Extract coordinates from part 1 result
    std::vector<Coord> trodden;
    trodden.reserve(pt1Path.size());
    for (const auto& entry : pt1Path) {
        trodden.push_back(entry.first);
    }
    
    // Calculate number of worker threads (use 75% of available cores)
    unsigned int numThreads = std::thread::hardware_concurrency() * 3 / 4;
    if (numThreads == 0) numThreads = 4; // Fallback
    
    // Thread-safe counter
    std::mutex countMutex;
    
    // Distribute work among threads
    std::vector<std::thread> threads;
    threads.reserve(numThreads);
    
    // Process chunk of coordinates per thread
    const int chunkSize = (trodden.size() + numThreads - 1) / numThreads;
    
    for (unsigned int t = 0; t < numThreads; t++) {
        threads.emplace_back([&, t]() {
            int localCount = 0;
            
            // Calculate range for this thread
            size_t start = t * chunkSize;
            size_t end = std::min(start + chunkSize, trodden.size());
            
            Grid localGrid = originalGrid; // Each thread gets its own grid copy
            
            for (size_t i = start; i < end; i++) {
                const auto& coord = trodden[i];
                
                // Skip guard starting position
                if (coord.row == startGuard.row && coord.col == startGuard.col) {
                    continue;
                }
                
                // Make a change to the grid without copying the whole thing
                char original = localGrid[coord.row][coord.col];
                localGrid[coord.row][coord.col] = '#';
                
                if (solve(localGrid).empty()) {
                    localCount++;
                }
                
                // Restore grid to original state for next iteration
                localGrid[coord.row][coord.col] = original;
            }
            
            // Accumulate local count to global count
            std::lock_guard<std::mutex> lock(countMutex);
            count += localCount;
        });
    }
    
    // Wait for all threads to complete
    for (auto& thread : threads) {
        thread.join();
    }
    
    return count;
}

int main() {
    auto start = std::chrono::high_resolution_clock::now();
    
    // Load grid once
    Grid originalGrid = getGrid();
    
    // Part 1
    auto path = runPt1(originalGrid);
    std::cout << "Part 1: " << path.size() << std::endl;
    
    // Part 2
    int part2 = runPt2(originalGrid, path);
    std::cout << "Part 2: " << part2 << std::endl;
    
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
    
    std::cout << "Time: " << duration << "ms" << std::endl;
    
    return 0;
} 

File: 2024/Day06_2.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Coord {
    public final int row, col;

    Coord(int row, int col) {
        this.row = row;
        this.col = col;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (!(o instanceof Coord))
            return false;
        Coord coord = (Coord) o;
        return row == coord.row && col == coord.col;
    }

    @Override
    public int hashCode() {
        return 31 * row + col;
    }
}

class Guard {
    public int[] pos;
    public String dir;

    Guard(int[] pos, String dir) {
        this.pos = pos;
        this.dir = dir;
    }
}

public class Day06_2 {
    public static final Map<String, int[]> nextPosMods = Map.of(
            "^", new int[] { -1, 0 },
            "v", new int[] { 1, 0 },
            "<", new int[] { 0, -1 },
            ">", new int[] { 0, 1 });
    public static final Map<String, String> turns = Map.of(
            "^", ">",
            "v", "<",
            "<", "^",
            ">", "v");
    public static final String filepath = "inputs/day06.txt";

    public static void main(String[] args) throws IOException {
        final long start = System.currentTimeMillis();

        var path = runPt1();
        List<Coord> keys = new ArrayList<Coord>(path.keySet());
        System.out.println(String.format("Part 1: %d", keys.size()));

        runPt2(keys);

        final long end = System.currentTimeMillis();
        System.out.println(String.format("Time: %dms", end - start));
    }

    public static Map<Coord, List<String>> runPt1() throws IOException {
        List<List<String>> grid = getGrid();
        return solve(grid);
    }

    public static void runPt2(List<Coord> trodden) throws IOException {
        List<List<String>> grid = getGrid();
        Guard guard = getGuard(grid);
        var count = 0;
        for (var coord : trodden) {
            if (coord.row == guard.pos[0] && coord.col == guard.pos[1]) {
                continue;
            }

            grid.get(coord.row).set(coord.col, "#");
            var completed = solve(grid);
            if (completed == null) {
                count++;
            }
            grid.get(coord.row).set(coord.col, ".");
        }
        System.out.println(String.format("Part 2: %d", count));
    }

    public static List<List<String>> getGrid() throws IOException {
        List<List<String>> grid = new ArrayList<>();
        BufferedReader br = new BufferedReader(new FileReader(filepath));
        String line;
        while ((line = br.readLine()) != null) {
            grid.add(new ArrayList<>(Arrays.asList(line.split(""))));
        }
        br.close();
        return grid;
    }

    public static Map<Coord, List<String>> solve(List<List<String>> grid) {
        Guard guard = getGuard(grid);
        Map<Coord, List<String>> path = new HashMap<>();
        path.put(new Coord(guard.pos[0], guard.pos[1]), new ArrayList<>(List.of("^")));
        while (isInGrid(grid, guard.pos)) {
            if (!tick(grid, guard, path)) {
                return null;
            }
        }
        return path;
    }

    public static Guard getGuard(List<List<String>> grid) {
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid.get(0).size(); j++) {
                if (grid.get(i).get(j).equals("^")) {
                    return new Guard(new int[] { i, j }, "^");
                }

            }
        }
        return null;
    }

    public static boolean isInGrid(List<List<String>> grid, int[] pos) {
        return pos[0] >= 0 && pos[0] < grid.size() && pos[1] >= 0 && pos[1] < grid.get(0).size();
    }

    public static boolean tick (List<List<String>> grid, Guard guard, Map<Coord, List<String>> path) {
        move(grid, guard);
        var pathKey = new Coord(guard.pos[0], guard.pos[1]);
        var val = path.get(pathKey);
        if (val != null && val.contains(guard.dir)) {
            return false;
        }
        if (isInGrid(grid, guard.pos)) {
            if (val == null) {
                path.put(pathKey, new ArrayList<>());
            }
            path.get(pathKey).add(guard.dir);
        }
        return true;
    }

    public static void move(List<List<String>> grid, Guard guard) {
        var nextPosMod = nextPosMods.get(guard.dir);
        var nextPos = new int[]{guard.pos[0] + nextPosMod[0], guard.pos[1] + nextPosMod[1]};
        if (isInGrid(grid, nextPos) && grid.get(nextPos[0]).get(nextPos[1]).equals("#")) {
            guard.dir = turns.get(guard.dir);
        } else {
            guard.pos = nextPos;
        }
    }
}
